ndonnx/_array.py

```diff
@@ -23,18 +23,86 @@
 from .extensions import get_mask
 from .types import GetItemKey, OnnxShape, PyScalar, SetitemKey
 
-_BinaryOp = Callable[["Array", "int | bool | str | float | Array"], "Array"]
+_BinaryOp = Callable[
+    ["Array", "PyScalar | Array | np.ndarray | np.generic"],
+    "Array",
+]
 _Axisparam = int | tuple[int, ...] | None
 
 
-def _make_binary(
+def _make_binary_dunder(
     tyarr_op: Callable[[TyArrayBase | PyScalar, TyArrayBase | PyScalar], TyArrayBase],
+    sigil: str,
 ) -> tuple[_BinaryOp, _BinaryOp]:
-    def binary_op_forward(self, other):
-        return _apply_op(self, other, tyarr_op)
-
-    def binary_op_backward(self, other):
-        return _apply_op(other, self, tyarr_op)
+    """Create a forward and reflected version for a binary dunder method."""
+
+    # If we return 'NotImplemented' from methods such as __add__ the
+    # interpreter will create an error message that does not display
+    # the arrays dtype. E.g. `"TypeError: ... +: Not Implemented for
+    # 'Array' and 'Other'`. This is ok, if `Other` is some unrelated
+    # class, but we would not want to show an error massage such as
+    # `"TypeError: ... +: Not Implemented for 'Array' and 'Array'` in
+    # cases where the data types are not compatible.
+    #
+    # We want to cover the following scenarios for incompatible operands.
+    #
+    # Other() + Array:
+    #   -> Array.__radd__:
+    #     -> return NotImplemented
+    # Array + Other():
+    #   -> Array.__add__:
+    #     -> return NotImplemented
+    # Array(dtype1) + Array(dtype2):
+    #   -> Option 1 (Taken in the current implementation):
+    #     -> operator.add(TyArray(dtype1), TyArray(dtype2)):
+    #        - Tries __add__ and __radd__ on the TyArray objects
+    #        - Raise TypeError with bad error message
+    #        - Catch and raise new error with better message
+    #   -> Option 2:
+    #     -> Manually try __add__ and __radd__ on the TyArray objects
+    #       -> Pass through the returned NotImplemented object
+    #       -> Still a bad error message for the user, but does not expose internal class names
+    #   -> Option 3:
```

Comment:
Can you please just check for `NotImplemented` on both implementations (i.e. option 3)? This try ... except logic feels brittle as all TypeErrors are now being trapped here and rewritten, which forces all the typed arrays to be extra careful when raising TypeErrors as they are all interpreted as being due to operand dtype mismatch. I'm especially concerned about custom dtypes as this behavior is undocumented.